### DISCUSSION TITLE
How to reload an SSR page from itself?

### DISCUSSION THREAD

**Discussion by theDanielJLewis** [vercel/next.js#51837] [Help] [answered]
### Summary

I built a dashboard page that displays multiple stats my app calculates for the user. But when they first add the item to the account, the stats will take a minute to collect and calculate, so I updated the dashboard page with skeletons until the item's status changes.

When the interval-based status check finds that the status has changed from `pending` to `active`, it needs to refresh the page.

But that's where I'm having problems. It seems the only way to refresh a page is with `useRouter()`, which can only be used in client components.

I tried using `redirect()`, but that throws errors and doesn't refresh the page.

Here's what my code looks like inside the async server component:

```js
// Get the item from the database
const item = await getOneItem(params.id);

// If the item status is pending, check the status every second and reload when it's active
if (item.status === 'pending') {
  async function checkStatus() {
    const podcast = await getOneitem(params.id, { status: 1 });
    console.log('status:', item.status);

    if (item.status === 'active') {
      clearInterval(refreshInterval);
      const random = Math.random();
      revalidatePath(`/items/${params.id}`);
      redirect(`/items/${params.id}?${random}`);
    }
  }

  const refreshInterval = setInterval(() => {
    checkStatus();
  }, 1000);
}
```

I've tried it with and without `revalidatePath()`, and with and without the random number in the redirect. When I use the redirect like this—no matter its destination, even if it make it only `/items`—I get this console error:

```
unhandledRejection: Error: NEXT_REDIRECT
    at getRedirectError (webpack-internal:///(sc_server)/./node_modules/next/dist/client/components/redirect.js:40:19)
    at redirect (webpack-internal:///(sc_server)/./node_modules/next/dist/client/components/redirect.js:50:11)
    at checkStatus (webpack-internal:///(sc_server)/./…/page.tsx:62:75)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  digest: 'NEXT_REDIRECT;replace;/items/6497cb1fb416c920776714a6?0.5812914171774763'
``` 

So how _should_ I be refreshing the current SSR page when the status changes to active?

Thanks in advance!

### Additional information

```js
Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:22 PDT 2023; root:xnu-8796.121.3~7/RELEASE_X86_64
    Binaries:
      Node: 18.15.0
      npm: 9.7.1
      Yarn: 1.22.19
      pnpm: 8.6.5
    Relevant packages:
      next: 13.4.7
      eslint-config-next: 13.4.7
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.3

warn  - Latest canary version not detected, detected: "13.4.7", newest: "13.4.8-canary.5".
        Please try the latest canary version (`npm install next@canary`) to confirm the issue still exists before creating a new issue.
        Read more - https://nextjs.org/docs/messages/opening-an-issue
```

**Comment by icyJoseph** [comment]
I think, the issue here, is that you already responded to the client. 

Why not await a promise instead? And use `loading.tsx` file? 

So when the page renders, it suspends, and the loading boundary is used instead. Then when you finally have data, you resolve the promise and render the page.

Though I think there's some issue with loading pages not working reliably after the first load.

**Reply by theDanielJLewis** [reply]
I apologize that I'm still new to this and a rather scrappy developer, so I don't completely understand your advice.

My understanding of `loading.tsx` is that it will display until the data is fetched. In my case, the DB record already exists, but it is missing fields required to display the stats on the dashboard. Thus, it would seem like a loading file would display only for the split second before the item is retrieved from the DB. If that's not the case, then I'm not sure how to trigger it with a promise like you seem to be suggesting.

**Reply by icyJoseph** [reply]
I'll have to re-think this through - it seems to be that I understood the use case totally wrong.

**Reply by theDanielJLewis** [reply]
Well, if there's a way I can make `loading.tsx` display _until_ my page sees the `completed` value in one of the item fields, that would be fine with me.

That's what I'm trying to do now, but I'm running into other problems because of the chain of dependencies other parts of the page has on that initial item load.

Here's my current attempt:

```js
	async function checkItem(itemId) {
		let item = {};

		const refreshInterval = setInterval(async () => {
			const tempItem = await getOneItem(itemId);

			console.log('status:', tempItem.status);

			if (tempItem.status === 'completed') {
				clearInterval(refreshInterval);
				item = tempItem;
			}
		}, 1000);

		return item as ItemClass;
	}

	const item = await checkItem(params.id);

```

**Reply by theDanielJLewis** [reply, marked as answer]
GOT IT!

> Why not await a promise instead? And use loading.tsx file?

That was the thing I had to keep thinking about. When I wrapped the inside `checkItem()` with `return new Promise(…)`, the page now displays the `loading.tsx` page until my function sees the expected status.

I don't really like that the browser loading indicator continues spinning during this time, but that's not a big deal.

Here's the code:

```js
async function checkItem(itemId): Promise<ItemClass> {
	return new Promise(async (resolve, reject) => {
		const item = await getOneItem(itemId);

		if (item.status === 'pending') {
			const refreshInterval = setInterval(async () => {
				const item = await getOneItem(itemId);
				console.log('status:', item.status);

				if (item.status === 'active') {
					clearInterval(refreshInterval);
					console.log('item is active now');
					resolve(item);
				}
			}, 1000);
		} else {
			resolve(item);
		}
	});
}

const item = await checkItem(params.id);
```

Thank you for pointing me in the right direction, @icyJoseph!

This didn't answer that question about reloads in server components, but at least it solved my problem.

**Reply by icyJoseph** [reply]
Awesome! Sorry I had a long day at work today, so I couldn't really check what the real use case was.

Yeah I think you kind of want to have a server action in your server component, and then pass that to a client component. 

And perhaps that client component calls the server action on intervals.

> Note that intervals tend to drift, but perhaps in your use case it is good enough:

```tsx
// app/page.tsx
import { revalidatePath } from "next/cache";
import { IntervalChecker } from "./IntervalChecker";

export default function Home() {
  const data = crypto.randomUUID();

  const shouldCheck = Math.random() > 0.5;

  async function check() {
    "use server";

    if (!shouldCheck) return;

    const rng = Math.random();

    if (rng > 0.5) {
      revalidatePath("/"); // or redirect back to itself 
      console.log("Done");
      return;
    }

    console.log("Retry");
  }

  return (
    <>
      <IntervalChecker enabled={shouldCheck} callback={check} />
      <div>{data}</div>
    </>
  );
}

```

With IntervalChecker:

```tsx
"use client";
// app/InteralChecker

import { startTransition, useEffect } from "react";

export const IntervalChecker = ({
  enabled,
  callback,
}: {
  enabled: boolean;
  callback: VoidFunction;
}) => {
  useEffect(() => {
    if (!enabled) return;

    let interval = setInterval(() => {
      startTransition(() => callback());
    }, 1_000);

    return () => {
      clearInterval(interval);
    };
  }, [enabled, callback]);

  return null;
};

```

In your case the `(item.status === 'pending')` should prevent further retries.

> Of course, activate the experimental server actions feature
